core/src/com/unciv/ui/screens/worldscreen/unit/UnitTable.kt

```diff
@@ -1,90 +1,101 @@
-    private val nextIdleUnitButton = IdleUnitButton(this,worldScreen.mapHolder,false, KeyboardBinding.NextIdleButton)
-    private val unitIconHolder = Table()
-    private val unitNameLabel = "".toLabel(fontSize = 24)
-    private val unitIconNameGroup = Table()
-    private val promotionsTable = Table().apply { defaults().padRight(5f) }
-    private val unitDescriptionTable = Table(BaseScreen.skin)
-    private val deselectUnitButton: Actor
-
-    val selectedUnit : MapUnit?
-        get() = selectedUnits.firstOrNull()
-    /** This is in preparation for multi-select and multi-move  */
-    val selectedUnits = ArrayList<MapUnit>()
-
-    // Whether the (first) selected unit is in unit-swapping mode
-    var selectedUnitIsSwapping = false
-
-    // Whether the (first) selected unit is in road-connecting mode
-    var selectedUnitIsConnectingRoad = false
+    internal val prevIdleUnitButton = IdleUnitButton(this, worldScreen.mapHolder, true, KeyboardBinding.PrevIdleButton)
+    internal val nextIdleUnitButton = IdleUnitButton(this, worldScreen.mapHolder, false, KeyboardBinding.NextIdleButton)
+    internal val unitIconHolder = Table()
+    internal val unitNameLabel = "".toLabel(fontSize = 24).apply { setAlignment(Label.CENTER) }
+    internal val unitIconNameGroup = Table()
+    internal val promotionsTable = Table().apply { defaults().padRight(5f) }
+    internal val descriptionTable = Table(BaseScreen.skin)
+    internal val closeButton: Actor
+    internal val separator: Actor
+
+    /**
+     * The unit table shows infos of selected units, cities, spies, and a summary if none of these
+     * are selected. Each one of them have their own presenters to show their data.
+     */
+    private var presenter: Presenter
+
+    private val unitPresenter = UnitPresenter(this, worldScreen)
+    private val cityPresenter = CityPresenter(this, unitPresenter)
+    private val spyPresenter = SpyPresenter(this)
+    private val summaryPresenter = SummaryPresenter(this)
 
-    /** Sending no unit clears the selected units entirely */
-    fun selectUnit(unit: MapUnit? = null, append: Boolean = false) {
-        if (!append) selectedUnits.clear()
-        selectedCity = null
-        if (unit != null) {
-            selectedUnits.add(unit)
-            unit.actionsOnDeselect()
-        }
-        selectedUnitIsSwapping = false
-        selectedUnitIsConnectingRoad = false
-        selectedSpy = null
-    }
-
-    var selectedCity : City? = null
 
     // This is so that not on every update(), we will update the unit table.
     // Most of the time it's the same unit with the same stats so why waste precious time?
-    var selectedUnitHasChanged = false
-    val separator: Actor
-
-    var selectedSpy: Spy? = null
-
-    fun selectSpy(spy: Spy?) {
-        selectedSpy = spy
-        selectedCity = null
-        selectedUnits.clear()
-        selectedUnitIsSwapping = false
-        selectedUnitIsConnectingRoad = false
-    }
+    var shouldUpdate = false
```

Comment:
@yairm210 do you think we could remove `shouldUpdate`, which was previously called `selectedUnitHasChanged`?

The in-code comment says:
```
// This is so that not on every update(), we will update the unit table.
// Most of the time it's the same unit with the same stats so why waste precious time?
```
This mechanism has almost no effect (anymore), because `UnitTable.update()` isn't called frequently. It would make the code a bit more concise, even though it currently also works with shouldUpdate in place.

Reply:
It's not? If the worldscreen updates and a unit is selected (which is most often the case) then the table would need to update as well, right? 

I'm not sure what the actual performance change is, but it's probably the for a reason? If you want to perf-test this be my guest, but 'simpler code at the cost of unknown performance degradation' is not a trade off I'm happy to make

Reply:
The `WorldScreen.shouldUpdate` works so efficiently though that UnitTable.update() gets called very rarely, only on user interaction. Anyway, let's keep the optimization in UnitTable, since it's already working :+1: